### PR TITLE
pshlogin: tests of login utility

### PR DESCRIPTION
**Pull request by niewim19**
Tests added for:
 - pshlogin (only ia32)
 - auth applet (all arch)

JIRA: PD-75

## Description
Two separate tests added as it is now impossible to properly test `pshlogin` on imxrt1064 (it needs binary rename).
 - `test-pshlogin.py` tests launching psh with login prompt. It does not test exitability as it needs launching it as a first psh. It is currently only for ia32
- `test-auth.py` is a similar test to `test-pshlogin.py` but it launches login prompt using `auth` command accessible in `psh`. It is a way to test login logic (searching for password etc) works without spawning new `psh`

As `pshlogin` uses `auth` applet not all features are duplicated in `pshlogin` tests, especially in places where there are no concerns of memory usage (e.g backspace handling)

Tests regarding issues with:
 - entering correct and incorrect combinations of login and password (together with illegal credentials)
 - handling entering and exiting login prompt
 - handling backspace

## Motivation and Context

Those two tests, with current state of repository, checks if everything works correctly with login related components of Phoenix-RTOS.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by jsarzy — CHANGES_REQUESTED** [review]
Overall it looks good, some remarks regarding to coding style.

Please split tools:
1. `tools.py` -> `tools/basic.py`
2. `tools_login.py` -> `tools/login.py`

It will require some changes in the other psh tests.
Also keep import in some order:
1. All standard library imports
2. External library imports
3. Our files

All these separated by a new line. It's suggestion but I think it will help us to keep imports readable

File: psh/test-auth.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import psh.tools_login as tools
+
+
+def assert_auth(p):
+    p.send('auth\n')
+    assert p.expect_exact(['auth', pexpect.TIMEOUT], timeout=1) == 0, 'Error on entering command auth'
+    assert p.expect_exact(['Login:', pexpect.TIMEOUT], timeout=2) == 0, '"auth" applet does not launch'
+
+
+def logging_in(p, login, passwd, iscorrect):
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/test-auth.py]: Wrapper removed. Thanks

File: psh/test-auth.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def harness(p):
+    # Run psh
+    run_psh(p)
+    assert_only_prompt(p)
+
+    # Check if auth app is available and login exit
+    assert_auth(p)
+    p.send('\x04')
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/test-auth.py]: Added `EOT` global

File: psh/test-auth.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    assert p.expect_exact(['(psh)%', pexpect.TIMEOUT], timeout=1) == 0, 'Cannot exit "auth" during login passing'
+
+    # Exiting auth durign password
+    assert_auth(p)
+    p.send(tools.cred_bad.login + '\n')
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/test-auth.py]: Credentials moved to tests

File: psh/test-auth.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    assert_auth(p)
+    p.send(printables + '\n')
+    assert p.expect_exact(['Login:', pexpect.TIMEOUT], timeout=2) == 0, 'Long login does not repeat login procedure'
+    logging_in(p, tools.cred_ok.login, printables, iscorrect=False)  # too long password
+
+    # Test backspace
+    assert_auth(p)
+    p.send(tools.cred_ok.login + tools.cred_ok.login)
+    for i in range(1, len(tools.cred_ok.login)+2):  # Test just enough backspaces
+        p.send('\x08')
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/test-auth.py]: Added `BACKSPACE` global. Thanks

File: psh/tools_login.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+from collections import namedtuple
+import pexpect
+
+Credentials = namedtuple('Credentials', 'login passwd')
+cred_ok = Credentials('defuser', '1234')
+cred_ok_etc = Credentials('root', '1234')
+cred_bad = Credentials('lorem', 'ipsum')
+cred_empty = Credentials('', '')
+
+psh_run_cmd = '/bin/pshlogin'
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/tools_login.py]: Good idea. Changed as requested.

File: psh/tools_login.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+
+
+from collections import namedtuple
+import pexpect
+
+Credentials = namedtuple('Credentials', 'login passwd')
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/tools_login.py]: Yea, this is good. Changed `login` to `user`. Thanks

File: psh/tools_login.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+def assert_failed_login(p):
+    assert p.expect_exact(['(psh)%', pexpect.TIMEOUT], timeout=1) > 0, 'Login should fail but passed'
+
+
+def assert_good_login(p):
+    assert p.expect_exact(['(psh)%', pexpect.TIMEOUT], timeout=1) == 0, 'Login should pass but failed'
+
+
+def login_procedure(p, login, passwd, iscorrect):
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/tools_login.py]: Function split is implemented. I just left three separate functions:
 - `assert_login`
 - 'assert_login_failed`
 - 'assert_login_empty'

two of them are somehow similar but i left them as it is not to introduce too much complexity (`assert_login` and `assert_login_failed` differ by only one line)

File: psh/tools_login.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    p.send(login + '\n')
+    if login:  # Two scenarios for empty and nonempty login
+        assert p.expect_exact([login, pexpect.TIMEOUT]) == 0, 'Cannot enter login to login prompt'
+        assert p.expect_exact(["Password:", pexpect.TIMEOUT]) == 0
+        p.send(passwd + '\n')
+        if iscorrect is True:
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/tools_login.py]: iscorrect was removed all together

File: psh/tools_login.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
+
+
+from collections import namedtuple
+import pexpect
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/tools_login.py]: implemented correct spaces between imports

File: psh/tools_login.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        p.send(passwd + '\n')
+        if iscorrect is True:
+            assert_good_login(p)
+        else:
+            assert_failed_login(p)
+    else:
+        assert p.expect_exact(['Login:', pexpect.TIMEOUT]) == 0, 'Empty login doesn`t repeat logging'
+
+
+def get_all_ascii():
```
[the remark this answers is not in the document]
Comment (niewim19) [reply psh/tools_login.py]: Thanks for that suggestion! Implemented

**Comment by niewim19** [comment]
@jsarzy Thanks for review. Beside code changes I also implemented a change in `psh` tests directory structure. `tools` directory was created and tool files `basics` and `login` reside there. Necessary changes were made to other `psh` tests.

File: psh/test-pshlogin.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+cred_ok = logintools.Credentials('defuser', '1234')
+cred_ok_etc = logintools.Credentials('root', '1234')
+cred_bad = logintools.Credentials('lorem', 'ipsum')
+cred_empty = logintools.Credentials('', '')
+
+
+def assert_pshlogin(p):
+    p.send(logintools.pshlogin_cmd + '\n')
+    p.expect_exact(logintools.pshlogin_cmd)
+    err = p.expect_exact(['Login:', '(psh)%', pexpect.TIMEOUT], timeout=1)
```
Comment (jsarzy) [inline psh/test-pshlogin.py:35]: Please use `idx` instead of `err`, is not some kind of error code

File: psh/test-pshlogin.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    if err == 1:
+        raise AssertionError('pshlogin is not available!')
+    if err == 2:
+        raise AssertionError('pshlogin does not launch')
+
+
+# def logging_in(p, login, passwd, iscorrect):
+#     # FIXME - when possible remove import function wrapper and use import directly
+#     logintools.login_procedure(p, login, passwd, iscorrect)
```
Comment (jsarzy) [inline psh/test-pshlogin.py:44]: Remove commented code

File: psh/test-pshlogin.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+    # Exit and try to login again but with defuser:
+    p.send(EOT)
+    assert_pshlogin(p)
+    logintools.assert_login_fail(p, cred_ok.user, cred_bad.passwd)
+    logintools.assert_login(p, cred_ok.user, cred_ok.passwd)
+    p.send(EOT)
+
+    # Too long credentials
+    printables = string.printable
```
Comment (jsarzy) [inline psh/test-pshlogin.py:76]: This variable is not needed, just use `string.printable`

File: psh/test-auth.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    assert_auth(p)
+    logintools.assert_login_fail(p, cred_ok.user, cred_empty.passwd)
+    assert_auth(p)
+    logintools.assert_login_fail(p, cred_bad.user, cred_empty.passwd)
+    assert_auth(p)
+    logintools.assert_login_empty(p, cred_empty.user)
+    p.send(EOT)  # logging_in() with empty login does not go back to psh, exit needed
+
+    # Too long credentials
+    printables = string.printable
```
Comment (jsarzy) [inline psh/test-auth.py:81]: this variable is not needed, just use string.printable

**Review by jsarzy — CHANGES_REQUESTED** [review]

**Comment by niewim19** [comment]
Summary comment to last code review.

Thank you for all notes! All were implemented as stated.

**Comment by niewim19** [comment]
Changes in good direction. Tests are now clearer. LGTM

**Review by jsarzy — APPROVED** [review]
LGTM

**Pull request merged**